src/gt4py/next/utils.py

```diff
@@ -66,8 +62,17 @@ def is_tuple_of(v: Any, t: type[_T]) -> TypeGuard[tuple[_T, ...]]:
     return isinstance(v, tuple) and all(isinstance(e, t) for e in v)
 
 
+# TODO(havogt): remove flatten duplications in the whole codebase
+def flatten_nested_tuple(value: tuple[_S | tuple, ...]) -> tuple[_S, ...]:
+    if isinstance(value, tuple):
+        return sum((flatten_nested_tuple(v) for v in value), start=())  # type: ignore[arg-type] # cannot properly express nesting
+    else:
+        return (value,)
```

Comment:
I believe your version is better because it is lazy, but the functionality is the same. Only in my version I am lying about the typing because otherwise you cannot pass a tuple because it immediately gets matched to `_S` and then you get a typing error. I opted for the version with a typing error inside but now obviously you cannot pass a value without getting a typing error. The best version uses a protocol `NestedSequence` that Enrique found in some codebase. I'll add an issue about this cleanup.